internal/datastore/proxy/replicated.go

```diff
@@ -0,0 +1,385 @@
+// not have the requested revision, the primary is used instead.
+//
+// Unlike NewCheckingReplicatedDatastore, this function does not check the replicas for the requested
+// revision before reading from them; instead, a revision check is inserted into the SQL for each read.
+// This is useful when the read pool points to a load balancer that can transparently handle the request.
+// In this case, the primary will be used as a fallback if the replica does not have the requested revision.
+// The replica(s) supplied to this proxy *must*, therefore, have strict read mode enabled, to ensure the
+// query will fail with a RevisionUnavailableError if the revision is not available.
+func NewStrictReplicatedDatastore(primary datastore.Datastore, replicas ...datastore.StrictReadDatastore) (datastore.Datastore, error) {
+	if len(replicas) == 0 {
+		log.Debug().Msg("No replicas provided, using primary as read source")
+		return primary, nil
+	}
+
+	cachingReplicas := make([]datastore.ReadOnlyDatastore, 0, len(replicas))
+	for _, replica := range replicas {
+		if !replica.IsStrictReadModeEnabled() {
+			return nil, fmt.Errorf("replica %v does not have strict read mode enabled", replica)
+		}
+
+		cachingReplicas = append(cachingReplicas, newCachedCheckRevision(replica))
+	}
+
+	log.Debug().Int("replica-count", len(replicas)).Msg("Using replicas for reads")
+	return &strictReplicatedDatastore{
+		primary,
+		cachingReplicas,
+		0,
+	}, nil
+}
+
+type checkingReplicatedDatastore struct {
+	datastore.Datastore
+	replicas []datastore.ReadOnlyDatastore
+
+	lastReplica uint64
+}
+
+// SnapshotReader creates a read-only handle that reads the datastore at the specified revision.
+// Any errors establishing the reader will be returned by subsequent calls.
+func (rd *checkingReplicatedDatastore) SnapshotReader(revision datastore.Revision) datastore.Reader {
+	var swapped bool
+	var next uint64
+	for !swapped {
+		last := rd.lastReplica
+		next = (rd.lastReplica + 1) % uint64(len(rd.replicas))
+		swapped = atomic.CompareAndSwapUint64(&rd.lastReplica, last, next)
+	}
+
+	log.Trace().Uint64("replica", next).Msg("choosing replica for read")
+	replica := rd.replicas[next]
+
+	return &checkingStableReader{
+		rev:     revision,
+		replica: replica,
+		primary: rd.Datastore,
+	}
+}
+
+type strictReplicatedDatastore struct {
+	datastore.Datastore
+	replicas []datastore.ReadOnlyDatastore
+
+	lastReplica uint64
+}
+
+// SnapshotReader creates a read-only handle that reads the datastore at the specified revision.
+// Any errors establishing the reader will be returned by subsequent calls.
+func (rd *strictReplicatedDatastore) SnapshotReader(revision datastore.Revision) datastore.Reader {
+	var swapped bool
+	var next uint64
+	for !swapped {
+		last := rd.lastReplica
+		next = (rd.lastReplica + 1) % uint64(len(rd.replicas))
+		swapped = atomic.CompareAndSwapUint64(&rd.lastReplica, last, next)
+	}
```

Comment:
We could probably simplify this whole thing if there is only one replica (likely the most common scenario) and avoid the superfluous compare and swap.

Reply:
Added a check